File: Leetcode/arrays/49. Group Anagrams/solution.cpp
// time complexity : O(n log n)
#include <vector>
#include<array>
#include <string>
#include <iostream>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        vector<vector<string>> res;

        unordered_map<string, vector<string>> mp;

        for (int i = 0; i < strs.size(); i++)
        {
            string temp = strs[i];
            sort(temp.begin(), temp.end());

            mp[temp].push_back(strs[i]);
        }

        for (auto i : mp)
        {
            res.push_back(i.second);
        }
        return res;
    }
};

int main()
{
    vector<string> input = {"eat", "tea", "tan", "ate", "nat", "bat"};
    Solution solution;

    vector<vector<string>> result;
    result = solution.groupAnagrams(input);

    for (const auto &group : result)
    {
        for (const string &str : group)
        {
            cout << str << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Leetcode/arrays/49. Group Anagrams/solution.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0)
            return new ArrayList<>();

        Map<String, List<String>> ansMap = new HashMap<>();
        int[] count = new int[26];

        for (String str : strs) {
            Arrays.fill(count, 0);

            for (char c : str.toCharArray()) {
                count[c - 'a']++;
            }

            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 26; i++) {
                sb.append('#');
                sb.append(count[i]);
            }

            String key = sb.toString();

            // System.out.println(key);
            if (!ansMap.containsKey(key)) {
                ansMap.put(key, new ArrayList<>());
            }
            ansMap.get(key).add(str);
        }

        return new ArrayList<>(ansMap.values());
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[] input = { "eat", "tea", "tan", "ate", "nat", "bat" };
        List<List<String>> result = sol.groupAnagrams(input);
        System.out.println(result);
        // Output: [[eat, tea, ate], [tan, nat], [bat]]
    }
}